chunky/src/java/se/llbit/chunky/resources/CompressedImage.java

```diff
@@ -0,0 +1,108 @@
+package se.llbit.chunky.resources;
+
+import it.unimi.dsi.fastutil.ints.Int2IntOpenHashMap;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+import se.llbit.math.QuickMath;
+
+/**
+ * Image using a palette compression to reduce memory usage.
+ * The size of the palette/number of bit each pixel takes
+ * is computed based on the number of colors used in the image to compress
+ */
+public class CompressedImage implements Image {
+  private int width;
+  private int height;
+  private int[] palette;
+  private int bitDepth;
+  private int pixelPerLong;
+  private long[] compressedData;
```

Comment:
is `int[]` without shifting logic faster? Textures aren't _that_ big in Minecraft (and not even with high-res resourcepacks)

Reply:
Even with int shifts would be need because each pixel can take an arbitrary number of bits (but for a given image this number is the same for all pixels).
Additionally a pixel is never stored between 2 variables, this means there are some wasted bits when the number of bit per pixel is not a divisor of 64 (or 32 if using int). For example is there is 7 bits per pixels, we can store 4 pixels in a int with 4 bits wasted, so 8 pixels are stored in 8 bytes. If using long, we can store 9 pixels with only 1 wasted bit. Biger types are better in this case.